_source/_posts/2022-04-26-webauthn-java.md

```diff
@@ -0,0 +1,713 @@
+
+Information is sent to the browser, where it's encoded and passed to the WebAuthn API as a [`PublicKeyCredential` object](https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredential). The response containing all the data needed for verification is sent to the server and used to create or verify a credential.
+
+#### Registration ceremony
+
+To kick off the registration ceremony, the client submits a uniquely identifying piece of information using a `fetch` request—in this case, a username. The server verifies that the username is unique and then creates and saves a 16-byte user handle. Configuration information about the server (like hostname and origin) was included in the `RelyingParty` bean when the server was created. This bean takes a `PublicKeyCredentialRequestOptions` object to collect the user information, and generates a challenge field, a random buffer of at least 16 bytes long. This object is cached on the server to preserve the challenge field, then turned into a JSON string and sent to the client.
+
+Once the information is received by the client, it's parsed into a JavaScript [`PublicKeyCredentialCreationOptions`](https://www.w3.org/TR/webauthn-2/#dictionary-makecredentialoptions) object. Then it is passed into the `navigator.credentials.create()` function as a `publicKey` option.
+
+> **IMPORTANT:** You will need to grab the client code from the example's [GitHub](https://github.com/oktadev/webauthn-java-example/tree/main/src/main/resources) repository.  Copy the `static` and `templates` folders to your project's `src/main/resources` directory.
+
+```javascript
+    .then(credentialCreateJson => ({
+        publicKey: {
+            ...credentialCreateJson.publicKey,
+            challenge: base64urlToUint8array(credentialCreateJson.publicKey.challenge),
+        user: {
+            ...credentialCreateJson.publicKey.user,
+            id: base64urlToUint8array(credentialCreateJson.publicKey.user.id),
+        },
+        excludeCredentials: credentialCreateJson.publicKey.excludeCredentials.map(credential => ({
+            ...credential,
+            id: base64urlToUint8array(credential.id),
+        })),
+        extensions: credentialCreateJson.publicKey.extensions,
+        },
+    }))
+    .then(credentialCreateOptions =>
+        navigator.credentials.create(credentialCreateOptions))
+```
+
+The browser checks with whatever source of authentication is available to it using the `Client to Authenticator Protocol` version two (CTAP2). This check can be implemented by the operating system, browser, or by a standalone authentication device connected by USB or Bluetooth. This device generates a series of fields—the most important are `credentialId` and `publicKey` (generated from a private key the authenticator should never share). Next, the browser collates this information into a `PublicKeyCredential`, which is turned into a JSON object, stringified, and sent back to the server.
```

Comment:
I'm not sure this should be wrapped in back ticks. It causes a weird break in the paragraph.